File: scripts/learning-map/collect-facets.py
#!/usr/bin/env python3

"""
collect-facets.py

Collects insights and creates learning map.
Output: Markdown for people + JSON for agents

Usage: python3 collect-facets.py <project-root>
"""

import hashlib
import json
import os
import sys
from datetime import datetime
from pathlib import Path

# Add current directory to path for config import
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from config import get_project_dir, get_vault_path


# --- Helpers ---
def get_project_id(project_root):
    return hashlib.md5(project_root.encode()).hexdigest()[:8]


def get_project_name(project_root):
    pkg_path = os.path.join(project_root, "package.json")
    if os.path.exists(pkg_path):
        with open(pkg_path, "r") as f:
            pkg = json.load(f)
            return pkg.get("name", os.path.basename(project_root))

    pyproject_path = os.path.join(project_root, "pyproject.toml")
    if os.path.exists(pyproject_path):
        with open(pyproject_path, "r") as f:
            content = f.read()
            import re

            match = re.search(r'name = "([^"]+)"', content)
            if match:
                return match.group(1)

    return os.path.basename(project_root)


def read_insight_index(vault_path):
    index_path = os.path.join(vault_path, "insight-index.jsonl")
    if not os.path.exists(index_path):
        return []

    insights = []
    with open(index_path, "r") as f:
        for line in f:
            line = line.strip()
            if line:
                try:
                    insights.append(json.loads(line))
                except json.JSONDecodeError:
                    pass
    return insights


def filter_insights_by_project(insights, project_root):
    return [i for i in insights if i.get("source", {}).get("project") == project_root]


def create_learning_map_markdown(insights, project_root, project_name):
    today = datetime.now().strftime("%Y-%m-%d")

    # Group insights by topic
    topics = {}
    for insight in insights:
        for text in insight.get("insights", []):
            words = " ".join(text.split()[:3])
            if words not in topics:
                topics[words] = []
            topics[words].append(
                {
                    "text": text,
                    "references": insight.get("references", []),
                    "timestamp": insight.get("recordedAt", ""),
                }
            )

    lines = []

    lines.append(f"# Learning Map: {project_name}")
    lines.append("")
    lines.append(f"**วันที่:** {today}")
    lines.append(f"**สรุป:** เรียนรู้ {len(insights)} insights จาก {len(topics)} topics")
    lines.append("")

    lines.append("## Topics")
    lines.append("")

    for i, (topic, items) in enumerate(topics.items(), 1):
        lines.append(f"### {i}. {topic}")
        lines.append("")

        for item in items:
            lines.append(f"- {item['text']}")
            if item["references"]:
                refs = ", ".join(
                    [f"`{r['file']}:{r['line']}`" for r in item["references"]]
                )
                lines.append(f"  - อ้างอิง: {refs}")

        lines.append("")

    lines.append("## Learning Path")
    lines.append("")
    lines.append("ลำดับการเรียนรู้:")
    lines.append("")

    for i, topic in enumerate(topics.keys(), 1):
        lines.append(f"{i}. **{topic}**")

    lines.append("")

    lines.append("## Summary")
    lines.append("")
    lines.append("Session นี้เรียนรู้:")
    lines.append(f"- {len(insights)} insights")
    lines.append(f"- {len(topics)} topics")
    lines.append("")

    return "\n".join(lines)


def create_learning_map_json(insights, project_root, project_name):
    project_id = get_project_id(project_root)
    today = datetime.now().strftime("%Y-%m-%d")

    return {
        "project": {"id": project_id, "name": project_name, "root": project_root},
        "session": {"date": today, "insight_count": len(insights)},
        "insights": insights,
    }


# --- Main ---
def main():
    if len(sys.argv) < 2:
        project_root = os.getcwd()
    else:
        project_root = sys.argv[1]

    if not os.path.exists(project_root):
        print(f"Project root does not exist: {project_root}")
        sys.exit(1)

    vault_path = get_vault_path()
    project_id = get_project_id(project_root)
    project_name = get_project_name(project_root)
    project_dir = get_project_dir(project_root)

    print(f"Vault: {vault_path}")
    print(f"Project: {project_name} ({project_id})")

    # Read insights
    all_insights = read_insight_index(vault_path)
    project_insights = filter_insights_by_project(all_insights, project_root)

    if not project_insights:
        print(f"No insights found for project: {project_name}")
        sys.exit(0)

    print(f"Found {len(project_insights)} insights")

    # Create project directory
    os.makedirs(project_dir, exist_ok=True)

    # Create Markdown (for people)
    markdown = create_learning_map_markdown(
        project_insights, project_root, project_name
    )
    md_path = os.path.join(project_dir, "learning-map.md")
    with open(md_path, "w") as f:
        f.write(markdown)
    print(f"Created: {md_path}")

    # Create JSON (for agents)
    json_data = create_learning_map_json(project_insights, project_root, project_name)
    json_path = os.path.join(project_dir, "learning-map.json")
    with open(json_path, "w") as f:
        json.dump(json_data, f, indent=2)
    print(f"Created: {json_path}")

    print("\nDone!")


if __name__ == "__main__":
    main()

File: scripts/learning-map/collect-facets.js
#!/usr/bin/env node

/**
 * collect-facets.js
 *
 * Collects insights and creates learning map.
 * Output: Markdown for people + JSON for agents
 *
 * Usage: node collect-facets.js <project-root>
 */

const fs = require("fs");
const path = require("path");
const crypto = require("crypto");
const os = require("os");
const { getVaultPath, getProjectDir, getSessionMapPath } = require("./config");

// --- Helpers ---
function getProjectId(projectRoot) {
	return crypto.createHash("md5").update(projectRoot).digest("hex").slice(0, 8);
}

function getProjectName(projectRoot) {
	try {
		const pkg = JSON.parse(
			fs.readFileSync(path.join(projectRoot, "package.json"), "utf8"),
		);
		return pkg.name || path.basename(projectRoot);
	} catch {
		try {
			const pyproject = fs.readFileSync(
				path.join(projectRoot, "pyproject.toml"),
				"utf8",
			);
			const match = pyproject.match(/name = "([^"]+)"/);
			return match ? match[1] : path.basename(projectRoot);
		} catch {
			return path.basename(projectRoot);
		}
	}
}

function readInsightIndex(vaultPath) {
	const indexPath = path.join(vaultPath, "insight-index.jsonl");
	if (!fs.existsSync(indexPath)) {
		return [];
	}

	const content = fs.readFileSync(indexPath, "utf8");
	return content
		.split("\n")
		.filter(Boolean)
		.map((line) => {
			try {
				return JSON.parse(line);
			} catch {
				return null;
			}
		})
		.filter(Boolean);
}

function filterInsightsByProject(insights, projectRoot) {
	return insights.filter((i) => {
		if (i.source && i.source.project) {
			return i.source.project === projectRoot;
		}
		return false;
	});
}

function createLearningMapMarkdown(insights, projectRoot, projectName) {
	const today = new Date().toISOString().split("T")[0];

	// Group insights by topic
	const topics = {};
	insights.forEach((insight) => {
		(insight.insights || []).forEach((text) => {
			const words = text.split().slice(0, 3).join(" ");
			if (!topics[words]) topics[words] = [];
			topics[words].push({
				text,
				references: insight.references || [],
				timestamp: insight.recordedAt,
			});
		});
	});

	// Build Markdown
	const lines = [];

	lines.push(`# Learning Map: ${projectName}`);
	lines.push("");
	lines.push(`**วันที่:** ${today}`);
	lines.push(
		`**สรุป:** เรียนรู้ ${insights.length} insights จาก ${Object.keys(topics).length} topics`,
	);
	lines.push("");

	lines.push("## Topics");
	lines.push("");

	Object.entries(topics).forEach(([topic, items], i) => {
		lines.push(`### ${i + 1}. ${topic}`);
		lines.push("");

		items.forEach((item) => {
			lines.push(`- ${item.text}`);
			if (item.references.length > 0) {
				const refs = item.references
					.map((r) => `\`${r.file}:${r.line}\``)
					.join(", ");
				lines.push(`  - อ้างอิง: ${refs}`);
			}
		});

		lines.push("");
	});

	lines.push("## Learning Path");
	lines.push("");
	lines.push("ลำดับการเรียนรู้:");
	lines.push("");

	Object.keys(topics).forEach((topic, i) => {
		lines.push(`${i + 1}. **${topic}**`);
	});

	lines.push("");

	lines.push("## Summary");
	lines.push("");
	lines.push("Session นี้เรียนรู้:");
	lines.push(`- ${insights.length} insights`);
	lines.push(`- ${Object.keys(topics).length} topics`);
	lines.push("");

	return lines.join("\n");
}

function createLearningMapJSON(insights, projectRoot, projectName) {
	const projectId = getProjectId(projectRoot);
	const today = new Date().toISOString().split("T")[0];

	// Convert insights to schema-compliant nodes
	const nodes = insights.map((insight) => {
		const text = (insight.insights || []).join(" ");
		const slug = text
			.toLowerCase()
			.replace(/[^a-z0-9]+/g, "-")
			.replace(/^-|-$/g, "")
			.slice(0, 60);

		return {
			id: `insight:${slug}`,
			type: "insight",
			name: text.slice(0, 100),
			content: text,
			tags: [],
			difficulty: "intermediate",
			source: insight.source
				? { type: "insight-index", id: insight.id }
				: undefined,
			created_at: insight.recordedAt || new Date().toISOString(),
		};
	}).filter((n) => n.content.length > 0);

	// Build edges from sequential insights (builds_on)
	const edges = [];
	for (let i = 1; i < nodes.length; i++) {
		edges.push({
			source: nodes[i - 1].id,
			target: nodes[i].id,
			type: "builds_on",
			description: `Sequential learning: ${nodes[i - 1].name.slice(0, 30)} → ${nodes[i].name.slice(0, 30)}`,
		});
	}

	return {
		version: "1.0.0",
		session: {
			date: today,
			summary: `เรียนรู้ ${nodes.length} insights`,
		},
		meta: {
			title: `Session ${today}`,
			locale: "th",
			quality_profile: "standard",
			views: [
				{ id: "all", label: "ทั้งหมด", focus: ["*"] },
				{ id: "flow", label: "ลำดับการเรียนรู้", focus: ["flow"] },
			],
		},
		nodes,
		edges,
	};
}

// --- Main ---
function main() {
	const projectRoot = process.argv[2] || process.cwd();
	const today = new Date().toISOString().split("T")[0];

	if (!fs.existsSync(projectRoot)) {
		console.error(`Project root does not exist: ${projectRoot}`);
		process.exit(1);
	}

	const vaultPath = getVaultPath();
	const projectId = getProjectId(projectRoot);
	const projectName = getProjectName(projectRoot);
	const projectDir = getProjectDir(projectRoot);

	console.log(`Vault: ${vaultPath}`);
	console.log(`Project: ${projectName} (${projectId})`);

	// Read insights
	const allInsights = readInsightIndex(vaultPath);
	const projectInsights = filterInsightsByProject(allInsights, projectRoot);

	if (projectInsights.length === 0) {
		console.log(`No insights found for project: ${projectName}`);
		process.exit(0);
	}

	console.log(`Found ${projectInsights.length} insights`);

	// Create project directory
	if (!fs.existsSync(projectDir)) {
		fs.mkdirSync(projectDir, { recursive: true });
	}

	// Create Markdown (for people)
	const markdown = createLearningMapMarkdown(
		projectInsights,
		projectRoot,
		projectName,
	);
	const mdPath = path.join(projectDir, "learning-map.md");
	fs.writeFileSync(mdPath, markdown);
	console.log(`Created: ${mdPath}`);

	// Create session map (schema-valid JSON)
	const json = createLearningMapJSON(projectInsights, projectRoot, projectName);
	const sessionMapDir = path.join(os.homedir(), ".learning-mode", "session-maps");
	if (!fs.existsSync(sessionMapDir)) {
		fs.mkdirSync(sessionMapDir, { recursive: true });
	}
	const jsonPath = path.join(sessionMapDir, `${projectId}-${today}.json`);
	fs.writeFileSync(jsonPath, JSON.stringify(json, null, 2));
	console.log(`Created: ${jsonPath}`);

	console.log("\nDone!");
}

if (require.main === module) {
	main();
}

module.exports = {
	getProjectId,
	getProjectName,
	createLearningMapMarkdown,
	createLearningMapJSON,
};
